lib/gems/pending/util/postgres_admin.rb

```diff
@@ -136,12 +136,24 @@ def self.backup_pg_compress(opts)
     opts.delete(:dbname)
 
     path = Pathname.new(opts.delete(:local_file))
-
     FileUtils.mkdir_p(path.dirname)
-    Dir.mktmpdir("vmdb_backup", path.dirname) do |dir|
-      runcmd("pg_basebackup", opts, :z => nil, :format => "t", :xlog_method => "fetch", :pgdata => dir)
-      FileUtils.mv(File.join(dir, "base.tar.gz"), path)
+
+    # Build commandline from AwesomeSpawn
+    args = {:z => nil, :format => "t", :xlog_method => "stream", :pgdata => "-"}
+    cmd  = AwesomeSpawn.build_command_line("pg_basebackup", combine_command_args(opts, args))
+
+    # Run command in a separate thread
+    read, write = IO.pipe
+    cmd_thread = Thread.new do
+      pid = Kernel.spawn(pg_env(opts), cmd, :out => write)
```

Comment:
This looks good, but I think it's a bit cleaner to use the thread for reading the output and let the "main" thread do all the process and thread waiting. How does this look?

```ruby
# spawn a subprocess to do the backup and a thread to read the output
backup_pid = Kernel.spawn(pg_env(opts), cmd, :out => write)
write.close
stream_reader = Thread.new { IO.copy_stream(read, path) }

# wait for them both to finish
Process.wait(backup_pid)
stream_reader.join
read.close

Reply:
Then you can also avoid the perfectly valid, but rather odd spawn from within a thread thing here.

Reply:
@carbonin hrmm.... now that I look at this closely, not sure this would work, at least in it's current form.

Part of what I was doing with the previous code was waiting until `Process.wait` was finished before calling `write.close`.  Here you just close it, which will break the pipe as you are trying to write to it from `pg_basebackup`, unless the backup is super quick (not likely).

I think moving `write.close` to under the `Process.wait(backup_pid)` will still allow this to work, and then the `stream_reader` should be able to `.join` shortly after.  I will test that out.

Reply:
Nope, pretty sure this works as that only closes the write end in the current process. Maybe if `spawn` doesn't actually create a separate process this would be a problem, but because this is essentially a `fork .. exec` this should work.

Actually I wrote a test program to be sure:
```ruby
r, w = IO.pipe
puts "forking writer"
pid = fork do
  begin
    10.times do
      sleep 1
      w.puts "things"
    end
  rescue => e
    puts "writer error #{e.message}"
  end
  puts "writer done"
end
w.close

puts "creating reader thread"
reader = Thread.new { IO.copy_stream(r, STDOUT) }

puts "waiting on writer"
Process.wait(pid)

puts "wiating on reader"
reader.join

r.close
```